File: ios-client/OutfitSuggestor/Utils/OutfitLayerExclusivity.swift
//
//  OutfitLayerExclusivity.swift
//  OutfitSuggestor
//
//  Jacket vs blazer mutual exclusivity — keep in sync with frontend outfitLayerExclusivity.ts
//

import Foundation

enum OutfitLayerExclusivity {
    private static let blazerPlaceholderPhrases = [
        "no structured blazer",
        "outfit built around your outerwear",
    ]

    struct LayerContext {
        var season: String? = nil
        var occasion: String? = nil
        var style: String? = nil
    }

    static func isWarmSeason(_ season: String?) -> Bool {
        let s = (season ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return s == "summer" || s == "warm"
    }

    static func prefersBlazerOverJacket(occasion: String?, style: String?) -> Bool {
        let occ = (occasion ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .replacingOccurrences(of: "_", with: "-")
        let sty = (style ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let blazerOccasions: Set<String> = [
            "work", "business", "formal", "office", "interview",
            "wedding", "wedding-guest", "date-night", "everyday",
        ]
        if blazerOccasions.contains(occ) { return true }
        if ["classic", "elegant", "formal", "business"].contains(sty) { return true }
        return false
    }

    static func shouldShowBlazerCard(suggestion: OutfitSuggestion) -> Bool {
        if shouldShowAnchoredOuterwearInCoreGrid(suggestion: suggestion) {
            return false
        }
        if OutfitItemCardSourceTag.resolvedUploadCategory(suggestion: suggestion) == "outerwear" {
            return false
        }
        return !isBlazerPlaceholder(suggestion.blazer)
    }

    static func resolveOuterwearDisplayText(
        suggestion: OutfitSuggestion,
        context: LayerContext = LayerContext()
    ) -> String? {
        let anchor = OutfitItemCardSourceTag.resolvedUploadCategory(suggestion: suggestion)
        if isWarmSeason(context.season), anchor != "outerwear" {
            return nil
        }
        if let outerwear = suggestion.outerwear?.trimmingCharacters(in: .whitespacesAndNewlines),
           !outerwear.isEmpty {
            let lower = outerwear.lowercased()
            if lower != "null" && lower != "none" && lower != "n/a" {
                if !isBlazerPlaceholder(suggestion.blazer) {
                    if prefersBlazerOverJacket(occasion: context.occasion, style: context.style) {
                        return nil
                    }
                    if anchor != "outerwear" {
                        return nil
                    }
                }
                return outerwear
            }
        }
        if anchor == "outerwear" {
            return "Your wardrobe jacket (uploaded item)"
        }
        return nil
    }

    static func shouldShowAnchoredOuterwearInCoreGrid(suggestion: OutfitSuggestion) -> Bool {
        OutfitItemCardSourceTag.resolvedUploadCategory(suggestion: suggestion) == "outerwear"
            && resolveOuterwearDisplayText(suggestion: suggestion) != nil
    }

    static func optionalLayerCategories(
        for suggestion: OutfitSuggestion,
        context: LayerContext = LayerContext()
    ) -> [String] {
        let anchor = OutfitItemCardSourceTag.resolvedUploadCategory(suggestion: suggestion)
        let all = ["sweater", "outerwear", "tie"]
        switch anchor {
        case "blazer":
            return ["tie"]
        case "outerwear":
            return ["tie"]
        default:
            if isWarmSeason(context.season) || shouldShowBlazerCard(suggestion: suggestion) {
                return all.filter { $0 != "outerwear" }
            }
            return all
        }
    }

    static func hasVisibleOptionalLayers(
        _ suggestion: OutfitSuggestion,
        context: LayerContext = LayerContext()
    ) -> Bool {
        !OutfitOptionalLayers.items(
            for: suggestion,
            allowedCategories: optionalLayerCategories(for: suggestion, context: context)
        ).isEmpty
    }

    private static func isBlazerPlaceholder(_ text: String) -> Bool {
        let lower = text.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if lower.isEmpty || lower == "null" || lower == "n/a" || lower == "none" {
            return true
        }
        if lower.hasPrefix("consider adding") { return true }
        return blazerPlaceholderPhrases.contains { lower.contains($0) }
    }
}

File: frontend/src/utils/outfitLayerExclusivity.ts
import type { OutfitSuggestion } from '../models/OutfitModels';
import { resolveUploadCategory, type OptionalOutfitCategoryKey } from './outfitItemThumbnail';

const BLAZER_PLACEHOLDER_PATTERNS = [
  'no structured blazer',
  'outfit built around your outerwear',
];

function isBlazerPlaceholder(text: string): boolean {
  const lower = text.trim().toLowerCase();
  if (!lower || lower === 'null' || lower === 'n/a' || lower === 'none') return true;
  if (lower.startsWith('consider adding')) return true;
  return BLAZER_PLACEHOLDER_PATTERNS.some((pattern) => lower.includes(pattern));
}

function isWarmSeason(season?: string | null): boolean {
  const s = (season || '').trim().toLowerCase();
  return s === 'summer' || s === 'warm';
}

function hasMeaningfulLayer(text?: string | null): boolean {
  if (text == null) return false;
  return !isBlazerPlaceholder(String(text));
}

/** Prefer blazer over jacket for work/classic/everyday looks. */
export function prefersBlazerOverJacket(opts?: {
  occasion?: string | null;
  style?: string | null;
}): boolean {
  const occ = (opts?.occasion || '').trim().toLowerCase().replace(/_/g, '-');
  const sty = (opts?.style || '').trim().toLowerCase();
  if (
    [
      'work',
      'business',
      'formal',
      'office',
      'interview',
      'wedding',
      'wedding-guest',
      'date-night',
      'everyday',
    ].includes(occ)
  ) {
    return true;
  }
  if (['classic', 'elegant', 'formal', 'business'].includes(sty)) return true;
  return false;
}

export type LayerDisplayOpts = {
  season?: string | null;
  occasion?: string | null;
  style?: string | null;
};

/** Text for the outerwear slot — never alongside a real blazer; null in summer. */
export function resolveOuterwearDisplayText(
  suggestion: OutfitSuggestion,
  sourceWardrobeCategory?: string | null,
  opts?: LayerDisplayOpts
): string | null {
  const anchor = resolveUploadCategory(suggestion, sourceWardrobeCategory);
  if (isWarmSeason(opts?.season) && anchor !== 'outerwear') {
    return null;
  }

  const raw = suggestion.outerwear;
  if (raw != null && String(raw).trim() !== '') {
    const lower = String(raw).trim().toLowerCase();
    if (lower !== 'null' && lower !== 'none' && lower !== 'n/a') {
      if (hasMeaningfulLayer(suggestion.blazer) && anchor !== 'outerwear') {
        return null;
      }
      if (
        hasMeaningfulLayer(suggestion.blazer) &&
        prefersBlazerOverJacket({ occasion: opts?.occasion, style: opts?.style })
      ) {
        return null;
      }
      return String(raw);
    }
  }
  if (suggestion.imageUrl && anchor === 'outerwear') {
    return 'Your wardrobe jacket (uploaded item)';
  }
  return null;
}

/** Jacket/coat upload replaces blazer in the main result grid. */
export function shouldShowAnchoredOuterwearInCoreGrid(
  suggestion: OutfitSuggestion,
  sourceWardrobeCategory?: string | null
): boolean {
  return (
    resolveUploadCategory(suggestion, sourceWardrobeCategory) === 'outerwear' &&
    resolveOuterwearDisplayText(suggestion, sourceWardrobeCategory) != null
  );
}

/** Upload anchors a casual jacket/coat — hide the structured blazer card. */
export function shouldShowBlazerCard(
  suggestion: OutfitSuggestion,
  sourceWardrobeCategory?: string | null
): boolean {
  if (shouldShowAnchoredOuterwearInCoreGrid(suggestion, sourceWardrobeCategory)) return false;
  if (resolveUploadCategory(suggestion, sourceWardrobeCategory) === 'outerwear') return false;
  return !isBlazerPlaceholder(suggestion.blazer ?? '');
}

/** Optional layers after upper-body exclusivity. */
export function optionalLayerCategories(
  suggestion: OutfitSuggestion,
  sourceWardrobeCategory?: string | null,
  opts?: LayerDisplayOpts
): OptionalOutfitCategoryKey[] {
  const anchor = resolveUploadCategory(suggestion, sourceWardrobeCategory);
  const categories: OptionalOutfitCategoryKey[] = ['sweater', 'outerwear', 'tie'];
  if (anchor === 'blazer') {
    return categories.filter((c) => c === 'tie');
  }
  if (anchor === 'outerwear') {
    return categories.filter((c) => c !== 'sweater' && c !== 'outerwear');
  }
  if (isWarmSeason(opts?.season) || shouldShowBlazerCard(suggestion, sourceWardrobeCategory)) {
    return categories.filter((c) => c !== 'outerwear');
  }
  return categories;
}

export function resolveOptionalLayerText(
  suggestion: OutfitSuggestion,
  key: OptionalOutfitCategoryKey,
  sourceWardrobeCategory?: string | null,
  opts?: LayerDisplayOpts
): string | null {
  if (key === 'outerwear') {
    return resolveOuterwearDisplayText(suggestion, sourceWardrobeCategory, opts);
  }
  const raw = suggestion[key];
  if (raw != null && String(raw).trim() !== '') {
    const lower = String(raw).trim().toLowerCase();
    if (lower !== 'null' && lower !== 'none' && lower !== 'n/a') {
      return String(raw);
    }
  }
  return null;
}

export function hasVisibleOptionalLayers(
  suggestion: OutfitSuggestion,
  sourceWardrobeCategory?: string | null,
  opts?: LayerDisplayOpts
): boolean {
  return optionalLayerCategories(suggestion, sourceWardrobeCategory, opts).some(
    (key) => resolveOptionalLayerText(suggestion, key, sourceWardrobeCategory, opts) != null
  );
}
